src/linker/Linker/SignatureGenerator.PartVisitor.cs

```diff
@@ -0,0 +1,207 @@
+			{
+				VisitTypeReference (method.DeclaringType, builder);
+				builder.Append ('.').Append (GetEscapedMetadataName (method));
+
+				if (method.HasGenericParameters)
+					builder.Append ("``").Append (method.GenericParameters.Count);
+
+				if (method.HasParameters || (method.CallingConvention == MethodCallingConvention.VarArg))
+					VisitParameters (method.Parameters, method.CallingConvention == MethodCallingConvention.VarArg, builder);
+
+				if (method.Name == "op_Implicit" || method.Name == "op_Explicit") {
+					builder.Append ('~');
+					VisitTypeReference (method.ReturnType, builder);
+				}
+			}
+
+			public void VisitProperty (PropertyDefinition property, StringBuilder builder)
+			{
+				VisitTypeReference (property.DeclaringType, builder);
+				builder.Append ('.').Append (GetEscapedMetadataName (property));
+
+				if (property.Parameters.Count > 0)
+					VisitParameters (property.Parameters, false, builder);
+			}
+
+			public void VisitEvent (EventDefinition evt, StringBuilder builder)
+			{
+				VisitTypeReference (evt.DeclaringType, builder);
+				builder.Append ('.').Append (GetEscapedMetadataName (evt));
+			}
+
+			public void VisitGenericParameter (GenericParameter genericParameter, StringBuilder builder)
+			{
+				Debug.Assert ((genericParameter.DeclaringMethod == null) != (genericParameter.DeclaringType == null));
+				// Is this a type parameter on a type?
+				if (genericParameter.DeclaringMethod != null) {
+					builder.Append ("``");
+				} else {
+					Debug.Assert (genericParameter.DeclaringType != null);
+
+					// If the containing type is nested within other types.
+					// e.g. A<T>.B<U>.M<V>(T t, U u, V v) should be M(`0, `1, ``0).
+					// Roslyn needs to add generic arities of parents, but the innermost type redeclares 
+					// all generic parameters so we don't need to add them.
+					builder.Append ('`');
+				}
+
+				builder.Append (genericParameter.Position);
+			}
+
+			public void VisitTypeReference (TypeReference typeReference, StringBuilder builder)
+			{
+				switch (typeReference) {
+				case ByReferenceType byReferenceType:
+					VisitByReferenceType (byReferenceType, builder);
+					return;
+				case PointerType pointerType:
+					VisitPointerType (pointerType, builder);
+					return;
+				case ArrayType arrayType:
+					VisitArrayType (arrayType, builder);
+					return;
+				case GenericParameter genericParameter:
+					VisitGenericParameter (genericParameter, builder);
+					return;
+				}
+
+				if (typeReference.IsNested) {
+					VisitTypeReference (typeReference.GetInflatedDeclaringType (), builder);
+					builder.Append ('.');
+				}
+
+				if (!String.IsNullOrEmpty (typeReference.Namespace))
+					builder.Append (typeReference.Namespace).Append ('.');
+
+				// This includes '`n' for mangled generic types
+				builder.Append (typeReference.Name);
+
+				if (typeReference.HasGenericParameters || !typeReference.IsGenericInstance)
```

Comment:
Is this condition correct?

Reply:
I think so - unless I missed a case in cecil. Non-generics don't need more work, and uninstantiated generics are already mangled. I added a clarifying comment.